### PR TITLE
Fix GCC warning: forbidden conversion to char*

### PR DESCRIPTION
**Pull request by plexoos**
GCC 11
warning: ISO C++ forbids converting a string constant to 'char*' [-Wwrite-strings]

GCC 485
warning: deprecated conversion from string constant to 'char*' [-Wwrite-strings]

See #392 for motivation

### CODE REVIEW

**Review by akioogawa — APPROVED** [review]

**Review by veprbl — APPROVED** [review]
Diff LGTM

**Comment by plexoos** [comment]
Thanks @veprbl @akioogawa It is still a WIP though.
With the fixes in this PR the number of the targeted warnings went down but there is still a lot more to take care of.

```diff
--- fix_warns_orig.out  2022-09-03 18:33:15.000000000 -0400
+++ fix_warns_pr.out    2022-09-03 18:33:47.000000000 -0400
@@ -1,6 +1,6 @@
                                                                       msg  count
-0         deprecated conversion from string constant to [-Wwrite-strings]    798
+0         deprecated conversion from string constant to [-Wwrite-strings]    439
 1      may be used uninitialized in this function [-Wmaybe-uninitialized]    425
 2   deleting object of polymorphic class type which has non-virtual de...    403
 3                   variable set but not used [-Wunused-but-set-variable]    341
 4                                     unused variable [-Wunused-variable]    293
```

**Review by kkauder — APPROVED** [review]

**Review by fgeurts — APPROVED** [review]

**Review by starsdong — APPROVED** [review]

**Review by Navagyan — APPROVED** [review]

File: StRoot/StStrangeMuDstMaker/StStrangeMuDstPlayer.cxx [hunk 1]
```diff
@@ -17,13 +19,13 @@ ClassImp(StrangeMuDstPlayer)
 
 char empty = '\0';
 char slash = '/';
-void ParseFileName(Char_t *output, Char_t **file, Char_t **dir) {
-  if ((*file = strrchr(output,slash))) {
+void ParseFileName(const Char_t *output, Char_t **file, Char_t **dir) {
```
Comment (veprbl) [inline StRoot/StStrangeMuDstMaker/StStrangeMuDstPlayer.cxx]:
```suggestion
void ParseFileName(const Char_t *output, const Char_t **file, Char_t **dir) {
```

Comment (veprbl) [reply StRoot/StStrangeMuDstMaker/StStrangeMuDstPlayer.cxx]: and s/`Char_t *file, *dir, *outfile[3], line[80];`/`const Char_t *file, *dir; Char_t *outfile[3], line[80];`/

Comment (plexoos) [reply StRoot/StStrangeMuDstMaker/StStrangeMuDstPlayer.cxx]: Looks like the output argument. You think it will compile?

Comment (veprbl) [reply StRoot/StStrangeMuDstMaker/StStrangeMuDstPlayer.cxx]: Compiles on godbolt https://godbolt.org/z/oax3Eqej9 and there is `const char` overloaded strrchr https://cplusplus.com/reference/cstring/strrchr/

Comment (veprbl) [reply StRoot/StStrangeMuDstMaker/StStrangeMuDstPlayer.cxx]: P.S. C pointer types are to be read right to left
https://cdecl.org/?q=declare+x+as+pointer+to+pointer+to+const+char -> `const char **x`
https://cdecl.org/?q=declare+x+as+pointer+to+const+pointer+to+char -> `char * const *x`

Comment (plexoos) [reply StRoot/StStrangeMuDstMaker/StStrangeMuDstPlayer.cxx]: I tried your suggestion but it did not work for me:

```
.sl79_gcc485/OBJ/StRoot/StStrangeMuDstMaker/StStrangeMuDstPlayer.cxx: In member function 'void StrangeMuDstPlayer::Make(Int_t, StFile*, const Char_t*)':
.sl79_gcc485/OBJ/StRoot/StStrangeMuDstMaker/StStrangeMuDstPlayer.cxx:117:38: error: invalid conversion from 'const Char_t** {aka const char**}' to 'Char_t** {aka char**}' [-fpermissive]
     ParseFileName(output, &file, &dir);
                                      ^
.sl79_gcc485/OBJ/StRoot/StStrangeMuDstMaker/StStrangeMuDstPlayer.cxx:22:6: error:   initializing argument 3 of 'void ParseFileName(const Char_t*, const Char_t**, Char_t**)' [-fpermissive]
 void ParseFileName(const Char_t *output, const Char_t **file, Char_t **dir) {
      ^
.sl79_gcc485/OBJ/StRoot/StStrangeMuDstMaker/StStrangeMuDstPlayer.cxx:121:23: error: invalid conversion from 'const Char_t* {aka const char*}' to 'char*' [-fpermissive]
       strcpy(outfile[i],dir);
                       ^
```

Comment (veprbl) [reply StRoot/StStrangeMuDstMaker/StStrangeMuDstPlayer.cxx]: @plexoos Could you also post your diff?

Comment (plexoos) [reply StRoot/StStrangeMuDstMaker/StStrangeMuDstPlayer.cxx]:
```diff
--- a/StRoot/StStrangeMuDstMaker/StStrangeMuDstPlayer.cxx
+++ b/StRoot/StStrangeMuDstMaker/StStrangeMuDstPlayer.cxx
@@ -19,7 +19,7 @@ ClassImp(StrangeMuDstPlayer)

 char empty = '\0';
 char slash = '/';
-void ParseFileName(const Char_t *output, Char_t **file, Char_t **dir) {
+void ParseFileName(const Char_t *output, const Char_t **file, Char_t **dir) {
   if ((*file = std::strrchr(const_cast<char*>(output),slash))) {
     (*file)++;
     *dir = new char[strlen(output)+5];
@@ -64,7 +64,7 @@ void StrangeMuDstPlayer::Make(Int_t NEvents, StFile* input, const Char_t* output
   StMcEventMaker *mcEventReader;
   StAssociationMaker *associator;
 //   StRandyTopMapMaker *topoMapFixer;  // RWitt: Commented 7/24/15 to remove dependency on StHbtMaker
-  Char_t *file, *dir, *outfile[3], line[80];
+  const Char_t *file, *dir; Char_t *outfile[3], line[80];
   TString prefix[3];
   Int_t mNDstMakers = 0;


```

Comment (veprbl) [reply StRoot/StStrangeMuDstMaker/StStrangeMuDstPlayer.cxx]: Okay, had to make dir constant too and massage it's creation (memory leaks preserved!).
```diff
diff --git a/StRoot/StStrangeMuDstMaker/StStrangeMuDstPlayer.cxx b/StRoot/StStrangeMuDstMaker/StStrangeMuDstPlayer.cxx
index 349fea08b7..a6a9c61dd7 100755
--- a/StRoot/StStrangeMuDstMaker/StStrangeMuDstPlayer.cxx
+++ b/StRoot/StStrangeMuDstMaker/StStrangeMuDstPlayer.cxx
@@ -19,13 +19,14 @@ ClassImp(StrangeMuDstPlayer)
 
 char empty = '\0';
 char slash = '/';
-void ParseFileName(const Char_t *output, Char_t **file, Char_t **dir) {
-  if ((*file = std::strrchr(const_cast<char*>(output),slash))) {
+void ParseFileName(const Char_t *output, const Char_t **file, const Char_t **dir) {
+  if ((*file = std::strrchr(output,slash))) {
     (*file)++;
-    *dir = new char[strlen(output)+5];
-    strncpy(*dir,output,(strlen(output)-strlen(*file)));
+    char *_dir = new char[strlen(output)+5];
+    strncpy(_dir,output,(strlen(output)-strlen(*file)));
+    *dir = _dir;
   } else {
-    *file = const_cast<char*>(output);
+    *file = output;
     *dir = &empty;
   }
 }
@@ -64,7 +65,7 @@ void StrangeMuDstPlayer::Make(Int_t NEvents, StFile* input, const Char_t* output
   StMcEventMaker *mcEventReader;
   StAssociationMaker *associator;
 //   StRandyTopMapMaker *topoMapFixer;  // RWitt: Commented 7/24/15 to remove dependency on StHbtMaker
-  Char_t *file, *dir, *outfile[3], line[80];
+  const Char_t *file, *dir; Char_t *outfile[3], line[80];
   TString prefix[3];
   Int_t mNDstMakers = 0;
 
@@ -175,7 +176,7 @@ void StrangeMuDstPlayer::Filter(Int_t NEvents, StFile* input, const Char_t* outp
   StStrangeMuDstMaker *v0MuDstMaker=0;
   StStrangeMuDstMaker *xiMuDstMaker=0;
   StStrangeMuDstMaker *kinkMuDstMaker=0;
-  Char_t *file, *dir, *outfile[3], line[80];
+  const Char_t *file, *dir; Char_t *outfile[3], line[80];
   TString prefix[3];
   Int_t mNDstMakers = 0;
 
@@ -380,7 +381,7 @@ void StrangeMuDstPlayer::Play(Int_t NEvents, StFile* input, const Char_t* output
   StMcEventMaker *mcEventReader;
   StAssociationMaker *associator;
   //   StRandyTopMapMaker *topoMapFixer; // RWitt: Commented 7/24/15 to remove dependency on StHbtMaker
-  Char_t *file, *dir, *outfile[3], line[80];
+  const Char_t *file, *dir; Char_t *outfile[3], line[80];
   TString prefix[3];
   Int_t mNDstMakers = 0;
 
@@ -601,7 +602,7 @@ void StrangeMuDstPlayer::Copy(Int_t NEvents, StFile* input, const Char_t* output
   StStrangeMuDstMaker *v0MuDstMaker=0;
   StStrangeMuDstMaker *xiMuDstMaker=0;
   StStrangeMuDstMaker *kinkMuDstMaker=0;
-  Char_t *file, *dir, *outfile[3], line[80];
+  const Char_t *file, *dir; Char_t *outfile[3], line[80];
   TString prefix[3];
   Int_t mNDstMakers = 0;
 
```

Comment (plexoos) [reply StRoot/StStrangeMuDstMaker/StStrangeMuDstPlayer.cxx]: Very nice! Much better than the const cast
Do you want to commit this patch?

Comment (veprbl) [reply StRoot/StStrangeMuDstMaker/StStrangeMuDstPlayer.cxx]: Thanks! Done

File: OnlTools/OnlinePlots/Infrastructure/ServerGui.cxx [hunk 2]
```diff
@@ -566,7 +566,8 @@ void ServerGui::DoFileButton()
 
 
   static TGFileInfo* fi = new TGFileInfo();
-  fi->fIniDir = "/a";
+  static const char* inidir = "/a";
+  fi->fIniDir = const_cast<char*>(inidir);
```
Comment (veprbl) [inline OnlTools/OnlinePlots/Infrastructure/ServerGui.cxx]:
```suggestion
  static char inidir[] = "/a";
  fi->fIniDir = inidir;
```

**Comment by plexoos** [comment]
At least these are gone

fix_warns_root5

```
 msg_id                                                                                                                                    msg
      0                                                                        deprecated conversion from string constant to [-Wwrite-strings]
...
```

fix_warns_root6

```
 msg_id                                                                                                                                    msg
...
     16                                                                                               defined but not used [-Wunused-variable]
...
     22                                                                                    too many arguments for format [-Wformat-extra-args]
...
```

**Review by veprbl — APPROVED** [review]
Good to merge

**Comment by plexoos** [comment]
@jml985 @tonko-lj I would like to hear your comments as most of the warnings come from your code in StRoot/RTS/ and OnlTools/

**Comment by jml985** [comment]
There are seem to be three areas here:

1.   OnlTools/OnlinePlots  --- Nobody cares about these.  These were the "Panitkin Plots" that have not been used since 2012 or so.
2.   OnlTools/Jevp   --- These are mostly my code but run under a close-to-standard offline environment and if they compile they are probably just fine.  I'd prefer if you just tell me what you want changed so I can do it, but updating the git directly probably won't cause any harm.
3.    StRoot/RTS    --- These are the DAQ code we share with offline.   Git is not the primary source, so changing it here is just a source of confusion.  The code has to work with old VXWORKS compilers so please contact me or Tonko if you want changes made here.

**Comment by plexoos** [comment]
From the private communication:

> On 9/21/22 10:32 AM, Landgraf, Jeffery M. wrote:
>
>  Sounds good.
>
> -Jeff
>
> > Hi Jeff,
> >
> > let's not step on each other so I propose I deal with
> > the DETs (e.g. the DAQ_XXX subdirs) & RTS_EXAMPLE directory,
> > none which should be used in vxWorks, right?
> >
> > And you deal with the rest (include files), for now at least?
> >
> > -- Tonko
> >
> >
> > > On Tue, Sep 20, 2022 at 6:44 PM Landgraf, Jeffery M. wrote:
> > >
> > > Ok, I'll check them.   Most likely I can do this friday.
> > >
> > > -Jeff
> > >
> > > > Hello,
> > > >
> > > > yeah, I'm also irritated and it is not that hard to change on my end.
> > > > I've been pruning my stuff with time. But I also don't know what happens
> > > > with vxWorks since I don't use it much anymore.
> > > >
> > > > Jeff, how about you change one/few file to "const char *" and see what happens with
> > > > vxWorks? I'm betting it will be fine but...
> > > >
> > > > -- Tonko

**Comment by plexoos** [comment]
Last call for the code owners to comment or approve

**Pull request merged**